src/GraphQL/Instrumentation/TelemetryProvider.cs

```diff
@@ -0,0 +1,139 @@
+#if NET5_0_OR_GREATER
+
+using System.Diagnostics;
+using GraphQL.DI;
+using GraphQL.Execution;
+using GraphQL.Types;
+using GraphQL.Validation;
+using GraphQLParser.AST;
+
+namespace GraphQL.Instrumentation;
+
+/// <summary>
+/// Provides telemetry through the <see cref="Activity">System.Diagnostics.Activity API</see> as an <see cref="IConfigureExecution"/> implementation.
+/// Derive from this class to add additional telemetry.
+/// <para>
+/// To use a derived class, call <see cref="GraphQLBuilderExtensions.ConfigureExecution{TConfigureExecution}(IGraphQLBuilder)"/> with the type of your derived class.
+/// </para>
+/// </summary>
+public class TelemetryProvider : IConfigureExecution
+{
+    private static readonly ActivitySource _activitySource;
+    private readonly TelemetryOptions _telemetryOptions;
+    private const string ACTIVITY_OPERATION_NAME = "graphql";
+
+    /// <summary>
+    /// Gets the source name used for telemetry.
+    /// </summary>
+    public static string SourceName => "GraphQL";
+
+    static TelemetryProvider()
+    {
+        _activitySource = new ActivitySource(SourceName, typeof(TelemetryProvider).Assembly.GetNuGetVersion());
+    }
+
+    /// <summary>
+    /// Initializes a new instance with the specified options.
+    /// </summary>
+    public TelemetryProvider(TelemetryOptions options)
+    {
+        _telemetryOptions = options;
+    }
+
+    /// <inheritdoc/>
+    public virtual float SortOrder => GraphQLBuilderExtensions.SORT_ORDER_CONFIGURATION;
+
+    /// <inheritdoc/>
+    public virtual async Task<ExecutionResult> ExecuteAsync(ExecutionOptions options, ExecutionDelegate next)
+    {
+        // start the Activity
+        using var activity = await StartActivity(options).ConfigureAwait(false);
+
+        // if no event listeners, do not record any telemetry
```

Comment:
See async/await state machine optimization in `InstrumentFieldsMiddleware`.

Reply:
I don't believe that applies here, as `await StartActivity` occurs prior to the `if` condition.

Reply:
(now renamed to `StartActivityAsync`)

Reply:
OK